File: src/regex/gv.zig
const std = @import("std");

// print any struct, that implements
// - fn getNode(self: Self, n: usize) ?struct { begin: bool, final: bool }
// - fn containsEdge(nfa: Self, a: usize, b: usize, c: u8) bool

pub fn print(g: anytype, w: anytype) !void {
    try w.print("digraph {{\n", .{});
    try w.print("  0 [shape=point];\n", .{});

    var nodes: usize = 0;
    while (g.getNode(nodes)) |n| {
        nodes += 1;
        if (n.begin) {
            try w.print("  0 -> {};\n", .{nodes});
        }
        try w.print("  {} [shape={s}];\n", .{ nodes, if (n.final) "doublecircle" else "circle" });
    }
    try w.print("\n", .{});

    for (0..nodes) |a| {
        for (0..nodes) |b| {
            for (0..256) |ch| {
                const c: u8 = @intCast(ch);
                if (g.containsEdge(a, b, c)) {
                    if (c >= ' ' and c < 127 and c != '"' and c != '\\') {
                        try w.print("  {} -> {} [label=\"{c}\"];\n", .{ a + 1, b + 1, c });
                    } else {
                        try w.print("  {} -> {} [label=\"\\\\{o}\"];\n", .{ a + 1, b + 1, c });
                    }
                }
            }
        }
    }
    try w.print("}}\n", .{});
}

File: src/re/gv.rs
/*
 * common interface for all automation
 * Graphviz(T) -- printable url link to graph visualizer
 */

pub struct AutomationInfo {
    nodes: usize,
    begin: usize,
    epsilon: Option<u8>,
}

pub trait Automation {
    fn info(&self) -> AutomationInfo;
    fn contains_edge(&self, a: usize, b: usize, c: u8) -> bool;
}

pub struct Graphviz<T>(T);
impl<T> fmt::Display for Graphviz<T>
where
    T: Automation,
{
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "https://dreampuf.github.io/GraphvizOnline/#digraph%7B")?;
        write!(f, "%0A%20%200%20%5Bshape%3D%22point%22%5D%3B")?;
        for a in 1..=self.0.info().nodes {
            write!(f, "%0A%20%20{}%20%5Bshape%3D%22circle%22%5D%3B", a)?;
        }
        write!(f, "%0A%20%200%2D%3E{}%3B", self.0.info().begin + 1)?;
        for a in 1..=self.0.info().nodes {
            for b in 1..=self.0.info().nodes {
                for c in 0..=255u8 {
                    if self.0.contains_edge(a - 1, b - 1, c) {
                        write!(f, "%0A%20%20{}%2D%3E{}%20%5Blabel%3D%22", a, b)?;
                        if c.is_ascii_digit() || c.is_ascii_alphabetic() {
                            write!(f, "{}", c as char)?;
                        } else if c >= b' ' && c <= b'~' {
                            write!(f, "%{:02X}", c)?;
                        } else {
                            write!(f, "%5C%5C{}", c)?;
                        }
                        write!(f, "%22%5D%3B")?;
                    }
                }
            }
        }
        write!(f, "%0A%7D")
    }
}
